File: contract/contract_test.go
package contract

import "testing"

func TestDecodeValidContract(t *testing.T) {
	data := []byte(`{
      "ir_version":"1.0",
      "contract":{"id":"orders","version":"1"},
      "input":{"kind":"record-stream"},
      "rules":[{"id":"id-required","kind":"record","path":"$.id","predicate":{"op":"required"},"on_breach":{"severity":"error","action":"reject"}}]
    }`)
	got, err := Decode(data)
	if err != nil {
		t.Fatalf("Decode() error = %v", err)
	}
	if got.Metadata.ID != "orders" || len(got.Rules) != 1 {
		t.Fatalf("unexpected contract: %+v", got)
	}
}

func TestDecodeRejectsInvalidIRVersion(t *testing.T) {
	data := []byte(`{"ir_version":"2.0","contract":{"id":"x","version":"1"},"input":{"kind":"record-stream"},"rules":[{"id":"r","kind":"record","path":"$.x","predicate":{"op":"required"},"on_breach":{"severity":"error","action":"reject"}}]}`)
	if _, err := Decode(data); err == nil {
		t.Fatal("Decode() accepted unsupported IR version")
	}
}

func TestDecodeRejectsInvalidRegex(t *testing.T) {
	data := []byte(`{"ir_version":"1.0","contract":{"id":"x","version":"1"},"input":{"kind":"record-stream"},"rules":[{"id":"r","kind":"record","path":"$.x","predicate":{"op":"regex","pattern":"["},"on_breach":{"severity":"error","action":"reject"}}]}`)
	if _, err := Decode(data); err == nil {
		t.Fatal("Decode() accepted invalid regex")
	}
}

func TestCanonicalDigestNormalizesDefaultsAndRuleOrder(t *testing.T) {
	first := []byte(`{"ir_version":"1.0","contract":{"id":"orders","version":"1"},"input":{"kind":"record-stream"},"rules":[{"id":"z-rule","kind":"record","path":"$.z","predicate":{"op":"required"},"on_breach":{"action":"reject"}},{"id":"a-rule","version":"1","kind":"record","path":"$.a","predicate":{"op":"required"},"on_breach":{"severity":"error","action":"reject"}}]}`)
	second := []byte(`{"rules":[{"on_breach":{"action":"reject","severity":"error"},"predicate":{"op":"required"},"path":"$.a","kind":"record","id":"a-rule","version":"1"},{"id":"z-rule","version":"1","kind":"record","path":"$.z","predicate":{"op":"required"},"on_breach":{"severity":"error","action":"reject"}}],"input":{"kind":"record-stream"},"contract":{"version":"1","id":"orders"},"ir_version":"1.0"}`)
	c1, err := Decode(first)
	if err != nil {
		t.Fatal(err)
	}
	c2, err := Decode(second)
	if err != nil {
		t.Fatal(err)
	}
	canonical, err := CanonicalJSON(c1)
	if err != nil {
		t.Fatal(err)
	}
	wantCanonical := `{"ir_version":"1.0","contract":{"id":"orders","version":"1"},"input":{"kind":"record-stream"},"rules":[{"id":"a-rule","version":"1","kind":"record","path":"$.a","predicate":{"op":"required"},"on_breach":{"severity":"error","action":"reject"}},{"id":"z-rule","version":"1","kind":"record","path":"$.z","predicate":{"op":"required"},"on_breach":{"severity":"error","action":"reject"}}]}`
	if string(canonical) != wantCanonical {
		t.Fatalf("canonical JSON = %s, want %s", canonical, wantCanonical)
	}
	digest1, err := Digest(c1)
	if err != nil {
		t.Fatal(err)
	}
	digest2, err := Digest(c2)
	if err != nil {
		t.Fatal(err)
	}
	if digest1 != digest2 || len(digest1) != len("sha256:")+64 {
		t.Fatalf("digests differ or have wrong format: %q and %q", digest1, digest2)
	}
}

func TestDecodeRejectsUnknownAndTrailingData(t *testing.T) {
	base := `{"ir_version":"1.0","contract":{"id":"x","version":"1"},"input":{"kind":"record-stream"},"rules":[{"id":"r","kind":"record","path":"$.x","predicate":{"op":"required"},"on_breach":{"action":"reject"}}]}`
	if _, err := Decode([]byte(`{"extra":true,` + base[1:])); err == nil {
		t.Fatal("Decode() accepted an unknown field")
	}
	if _, err := Decode([]byte(base + base)); err == nil {
		t.Fatal("Decode() accepted trailing JSON")
	}
}

File: sdk/typescript/test/contract.test.ts
import assert from 'node:assert/strict'
import { test } from 'node:test'
import { Contract, Rule, predicate } from '../src/index.js'

test('matches the Go engine canonical digest', () => {
  const contract = new Contract('orders', '1', [
    new Rule('id-required', '$.id', predicate('required'), 'reject'),
    new Rule('email-valid', '$.email', predicate('format', { format: 'email' }), 'quarantine'),
    new Rule('amount-positive', '$.amount', predicate('range', { min: 0 }), 'reject'),
  ])
  assert.equal(contract.digest(), 'sha256:cd87b5f698ac1503f6eb622e5f560425e51260933b3be5d09e83e47dddd70503')
})

test('derives rule kind and validates stream constraints', () => {
  assert.equal(new Rule('unique-id', '$.id', predicate('unique'), 'reject').toIR().kind, 'stream')
  assert.throws(() => new Rule('unique-id', '$.id', predicate('unique'), 'reject', { kind: 'record' }).toIR(), /must be a stream rule/)
  assert.throws(() => new Rule('count', '$.id', predicate('count', { min: 1 }), 'reject').toIR(), /count path must be \$/)
})

test('normalizes defaults and sorts rules', () => {
  const contract = new Contract('x', '1', [
    new Rule('z', '$.z', predicate('required'), 'reject'),
    new Rule('a', '$.a', predicate('required'), 'reject'),
  ])
  const ir = contract.toIR()
  assert.deepEqual(ir.rules.map((rule) => rule.id), ['a', 'z'])
  assert.equal(ir.rules[0].version, '1')
  assert.equal(ir.rules[0].on_breach.severity, 'error')
})
